### PR TITLE
Proper docker files (build from code) - 2.1

### PR DESCRIPTION
**Pull request by totti4ever**
After a technical surprise which saw my changes disappear from my repo and mergify merging the empty pull request into this repo, this is hopefully the last try on this (as I said: the first PR I'm doing :-) )

Idea (and code) is the same as #305 and #288:

* Base image with all binaries (make, make-zip, make-tools)
* Dependant image for all three sub-images (restserver, joex, consumedir)
* build scripts reduced to one for building and one for pushing - they derive themselves whether it is a snapshot or a tagged version
* plus I added (basic) health checks

----

Issues to solve:
- [x] go back to local code instead of cloning git
may as second (but default) way of building. The current way might help to remotely build without having the code cloned
- [x] revert renaming of the docker images to one repo _docspell_ and having the separate images as version e.g. _docspell:consumedir-latest_
- [x] make UNOCONV URL overwirtable as argument as well
- improve caching during build
	- ~~use workaround in the cloning world to make sure a new commit leads to a new image in the latest world~~
_right now we don't support remote builds, which means everybody who wants to build, has to clone the repo. This won't work with DockerHub building the images itself, but since we push them ourselves that doesn't really matter_
	- [x] make sure that apk installs are cached and executed before git code is copied into the image
this could also mean going back to a base image at least for joex
	- [x] maybe having different base images any way to be able to build only one of the components, if the others are unchanged

### CODE REVIEW

**Comment by totti4ever** [comment]
@eikek, besides your performance thought, where I would need more input, I'd say I'm done with my tasks and you give this PR a test again! :-)

**Comment by totti4ever** [comment]
@eikek, I'm done I hope

Still working on the DockerHub integration, but will do that in a different PR

File: docker/hooks/build [hunk 1]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+./dev-build-images.sh "totti4ever/"
```
Comment (eikek) [inline docker/hooks/build:3]: Is this on purpose here? Or is it already for the docker-hub integration?

**Comment by eikek** [comment]
Thanks a lot again! I really really like it!

I have only one little thing that you couldn't know: the build requires (not a hard requirement) the `.git` directory to exist, because it is used to get the HEAD commit hash to include it into the application. This is displayed in the version information. This is not really necessary, but it looks missing in the webui. It would be nice, if this could be fixed (i think `.git` needs to be removed from .dockerignore?). If not, it's not a blocker.

Another very very minor thing :) Would it make sense to log to the console by default? I can always pipe it into some file if necessary… but I might miss something here (and it's not important).

File: docker/hooks/build [hunk 2]
```diff
@@ -11,6 +11,10 @@ echo "||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
 
 ./dev-build-images.sh "$DOCKER_REPO/"
 
+echo "#### pushing images ####"
+./dev-build-images.sh "$DOCKER_REPO/"
```
Comment (eikek) [inline docker/hooks/build:15]: Is this a typo? Shouldn't this  be `dev-push-images`?

File: docker/hooks/build [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+echo && echo && echo
+echo "||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||"
+
+
+./dev-build-images.sh "$DOCKER_REPO/"
+status=$?
+
+if [[ $status -eq 0 ]]; then
+  echo "#### pushing images ####"
+  ./dev-build-images.sh "$DOCKER_REPO/"
```
Comment (eikek) [inline docker/hooks/build:22]: Isn't this the same commad as above? Did you maybe meant to write `dev-push-images.sh` instead? (sorry if i don't get it…)

Comment (totti4ever) [reply docker/hooks/build]: you got it completely right, fixed ;-)

File: docker/hooks/build (hunk 2 above)
Comment (totti4ever) [reply docker/hooks/build:17]: same comment as below, isn't it? Should be fixed, thanks!

File: docker/hooks/build (hunk 1 above)
Comment (totti4ever) [reply docker/hooks/build]: Was is testing phase for the docker hub integration, indeed. Is replaced with a generic approach nevertheless now

**Comment by totti4ever** [comment]
@eikek, test run was successful

Please give it a last review and have a test yourself

thanks a lot

P.S.: .git is added again and log to console is default

**Comment by eikek** [comment]
Thank you so much! Just ran the script and started using docker-compose and everything works fine.

File: version.sbt [hunk 4]
```diff
@@ -1 +1 @@
-version in ThisBuild := "0.28.1"
+version = "0.13.0-SNAPSHOT"
```
Comment (eikek) [inline version.sbt]: It should work like this currently, but it's a bit safer to keep that `in ThisBuild`.

Comment (totti4ever) [reply version.sbt]: That was an incomplete revert, sorry! With https://github.com/totti4ever/docspell/commit/8d3549e4022f6f7dbc829bf5129de77623c0a56a it should be correct again now (never intended to change anything) and I cannot see it under changes in this PR anymore

Comment (eikek) [reply version.sbt]: ah ok

**Comment by totti4ever** [comment]
Sorry, I had to implement #312  together with the docker image restructuring, because it is heavily dependant on this PR as it basically further restructures the docker images:

* we have several base-images:
** `base-binaries` which builds the binaries which then later are copied from here to the other images
** `base` replaces the `base` we had before in that the other images are actually building on top of this. This is convenient for adding the maintainer, but is necessary especially for setting the DB variables at a central point.

So, in the end, what changes?
* there is an `.env` file which holds central parameters:
** the integration key
** the DB connection params like host, password etc
* if you don't have it, everything is still defaulted to the basic configuration and fits example docker-compose file with postgres

**Pull request merged**